tests/test_io.py

```diff
@@ -268,6 +268,37 @@ def test_list_dir_names_errors(tmpdir):
         ([], '.', '', [], True, False),
         ([], '.', 'wav', [], False, False),
         ([], '.', 'wav', [], True, False),
+        # file
+        (
+            ['file.txt'],
+            'file.txt',
+            '',
+            ['file.txt'],
+            False,
+            False,
+        ),
+        pytest.param(
+            [],
+            'file',
```

Comment:
Could you also add a test for `path='./file'`, I would expect this should return `[]` then.

Reply:
I think this part is still a little bit confusing.
In the docstring we state:

![image](https://user-images.githubusercontent.com/173624/216592544-bd0e2465-24f6-4590-aa35-26a77df4883f.png)

I would then interpret `file.txt` as a file matching pattern, otherwise you shouldn't also not be able to get more than one entry returned in the list.
The problem is when the user now presents just a single part of a path like `'file'` or `'t1.wav'` is this then interpreted as a path to a directory and we raise an error if the directory does not exist or should we interpret as a filename pattern?

Reply:
Your current solution seems to handle it like this:
if `path` does not exists as a file or directory and does not include any chars indicating it to be a pattern like `*`, `?`, ... it will interpret the path as a directory and not a pattern and raise an error as the folder does not exist.

If we stay with that solution, we need to update the docstring accordingly.

An alternative solution would be to treat the last part of the path (it's basename) always as search pattern as long as it don't end with `/` as you proposed in 1. in https://github.com/audeering/audeer/issues/76#issuecomment-1354558815.

Reply:
> An alternative solution would be to treat the last part of the path (it's basename) always as search pattern as long as it don't end with / as you proposed in 1

What is problematic with this solution is:

``` python
>>> audeer.path('dir/')
/dir
```

Of course, the better interface would have been to have separate arguments for the directory and an optional pattern. But now we are stuck with a single argument, so we need to figure out if `path` ends on a pattern or not.

Reply:
Yes, we need to stick with the single argument. When we don't want to use `/` at the end, which also seems risky to me, there are the following alternatives:

* every time path includes a `/` the second part is treated as a pattern, which will have the side effect that `'file'` and `'./file'` might return different results
* implement it as `ls` and raise also an error if no file is matched by the pattern
* only raise an error if `os.path.dirname(path)` does not exist and return empty list in all other none-matched, none-existing cases

Reply:
> every time path includes a / the second part is treated as a pattern, which will have the side effect that 'file' and './file' might return different results

If I understand you correctly, this would mean we generally remove support for listing all files in a folder, which seems to me the most common use-case of the function.

> implement it as ls and raise also an error if no file is matched by the pattern

This would mean we get an empty list for `list_file_names('/empty/folder')` but an error with `list_file_names(''/empty/folder/*')`.

> only raise an error if `os.path.dirname(path)` does not exist and return empty list in all other none-matched, none-existing cases

Not sure, but I think this is what we already do at the moment. So we would not solve #67.

Actually I have a fourth suggestion and that is we leave `list_file_names()` as is, but deprecate it in favor of a new function where we separate folder and pattern, e.g. `list_files()`. We can then also get rid of `filetype`, which is not really needed since a user can do `*.<ext>` and if we want to change the default parameter of `hidden` to `True` as we do with `list_dir_names()`.

Reply:
> Having an additional pattern variable feels also a little bit strange as ls just works in bash.

Yes, but also only with the kind of strange side-effect that it raises an error if the directory exists but no files match, which I don't like.

Otherwise, I'm also fine with the solution proposed here. The only downside I see is that `list_file_names('./file.txt', recursive=True)` will raise an error if `./file.txt` does not exist but only `./sub/file.txt`. In that case I would expect it returns `['./sub/file.txt']`, but since `./file.txt` is not recognizable as a pattern it will treat it as a directory and raise an error.

Reply:
I proposed changes to the docstring to handle the `file.txt` does not exists case, and agree with you that we should still include the changes to `audeer.list_file_names()`.

We can then still decide if we need an additional `audeer.list_files()` solution.

Reply:
We now have a solution for the missing case, see https://github.com/audeering/audeer/pull/94#discussion_r1098763187, so I guess there is not really a need for introducing a new function.